io.servicetalk.http.router.jersey.CancellationTest.cancelSse

Failed on CI: https://ci.servicetalk.io/job/servicetalk-master-prb/1131

```
Error Message
java.lang.AssertionError
Stacktrace
java.lang.AssertionError
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter$ConnectedPublisher.wakeupWriterThread(ConnectablePayloadWriter.java:340)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter$ConnectedPublisher.terminateRequestN(ConnectablePayloadWriter.java:330)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter$ConnectedPublisher.access$700(ConnectablePayloadWriter.java:245)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter$ConnectedPublisher$1.cancel(ConnectablePayloadWriter.java:294)
	at io.servicetalk.concurrent.api.DoSubscriptionPublisher$DoSubscriptionPublisherSubscriber$BeforeSubscription.cancel(DoSubscriptionPublisher.java:94)
	at io.servicetalk.http.api.HttpDataSourceTranformations$BridgeFlowControlAndDiscardSubscriber$2.cancel(HttpDataSourceTranformations.java:131)
	at io.servicetalk.concurrent.api.DoSubscriptionPublisher$DoSubscriptionPublisherSubscriber$BeforeSubscription.cancel(DoSubscriptionPublisher.java:94)
	at io.servicetalk.concurrent.api.ContextPreservingSubscription.cancel(ContextPreservingSubscription.java:54)
	at io.servicetalk.concurrent.internal.ConcurrentSubscription.cancel(ConcurrentSubscription.java:141)
	at io.servicetalk.concurrent.api.ContextPreservingCancellable.cancel(ContextPreservingCancellable.java:45)
	at io.servicetalk.concurrent.internal.SequentialCancellable.cancel(SequentialCancellable.java:69)
	at io.servicetalk.http.router.jersey.CancellationTest.testCancelResponsePayload(CancellationTest.java:205)
	at io.servicetalk.http.router.jersey.CancellationTest.cancelSse(CancellationTest.java:181)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.mockito.internal.junit.JUnitRule$1.evaluateSafely(JUnitRule.java:52)
	at org.mockito.internal.junit.JUnitRule$1.evaluate(JUnitRule.java:43)
	at io.servicetalk.concurrent.internal.ServiceTalkTestTimeout$TimeoutStatement$CallableStatement.call(ServiceTalkTestTimeout.java:167)
	at io.servicetalk.concurrent.internal.ServiceTalkTestTimeout$TimeoutStatement$CallableStatement.call(ServiceTalkTestTimeout.java:159)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Standard Output
[MockitoHint] CancellationTest.cancelSingle (see javadoc for MockitoHint):
[MockitoHint] 1. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:116)
[MockitoHint] 2. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:117)
[MockitoHint] 3. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:118)
[MockitoHint] 4. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:119)
[MockitoHint] 5. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:120)

[MockitoHint] CancellationTest.cancelOioStreams (see javadoc for MockitoHint):
[MockitoHint] 1. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:119)
[MockitoHint] 2. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:120)

[MockitoHint] CancellationTest.cancelSuspended (see javadoc for MockitoHint):
[MockitoHint] 1. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:118)
[MockitoHint] 2. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:120)

[MockitoHint] CancellationTest.cancelRsStreams (see javadoc for MockitoHint):
[MockitoHint] 1. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:119)
[MockitoHint] 2. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:120)

[MockitoHint] CancellationTest.cancelOffload (see javadoc for MockitoHint):
[MockitoHint] 1. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:116)
[MockitoHint] 2. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:117)
[MockitoHint] 3. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:118)
[MockitoHint] 4. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:119)
[MockitoHint] 5. Unused -> at io.servicetalk.http.router.jersey.CancellationTest.setup(CancellationTest.java:120)

Standard Error
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.server.ServerRuntime$Responder writeResponse
SEVERE: An I/O error has occurred while writing a response message entity to the container output stream.
org.glassfish.jersey.server.internal.process.MappableException: java.io.IOException: Already closed TerminalNotification{COMPLETE}
	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:67)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1111)
	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:638)
	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:371)
	at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:361)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:679)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter.handle0(DefaultJerseyStreamingHttpRouter.java:231)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter.access$000(DefaultJerseyStreamingHttpRouter.java:59)
	at io.servicetalk.http.router.jersey.DefaultJerseyStreamingHttpRouter$3.handleSubscribe(DefaultJerseyStreamingHttpRouter.java:169)
	at io.servicetalk.concurrent.api.Single.safeHandleSubscribe(Single.java:1538)
	at io.servicetalk.concurrent.api.ContextPreservingConsumer.accept(ContextPreservingConsumer.java:37)
	at io.servicetalk.concurrent.api.NoopOffloader.offloadSubscribe(NoopOffloader.java:74)
	at io.servicetalk.concurrent.api.Single.handleSubscribe(Single.java:1533)
	at io.servicetalk.concurrent.api.Single.subscribeWithOffloaderAndContext(Single.java:1504)
	at io.servicetalk.concurrent.api.Single.subscribeWithContext(Single.java:1485)
	at io.servicetalk.concurrent.api.Single.subscribeCaptureContext(Single.java:1457)
	at io.servicetalk.concurrent.api.Single.subscribeInternal(Single.java:1039)
	at io.servicetalk.concurrent.api.SingleFlatMapSingle$SubscriberImpl.onSuccess(SingleFlatMapSingle.java:83)
	at io.servicetalk.concurrent.api.ContextPreservingCancellableSingleSubscriber.onSuccess(ContextPreservingCancellableSingleSubscriber.java:42)
	at io.servicetalk.concurrent.api.ContextPreservingSingleSubscriber.onSuccess(ContextPreservingSingleSubscriber.java:51)
	at io.servicetalk.concurrent.api.AbstractSubmitSingle.lambda$handleSubscribe$0(AbstractSubmitSingle.java:49)
	at io.servicetalk.concurrent.api.ContextPreservingRunnable.run(ContextPreservingRunnable.java:39)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Already closed TerminalNotification{COMPLETE}
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.processClosed(ConnectablePayloadWriter.java:174)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.processClosed(ConnectablePayloadWriter.java:166)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.waitForRequestNDemand(ConnectablePayloadWriter.java:199)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.write(ConnectablePayloadWriter.java:93)
	at io.servicetalk.concurrent.api.internal.ConnectableOutputStream.write(ConnectableOutputStream.java:35)
	at io.servicetalk.http.router.jersey.DefaultContainerResponseWriter$CopyingOutputStream.write(DefaultContainerResponseWriter.java:296)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.write(CommittingOutputStream.java:219)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$UnCloseableOutputStream.write(WriterInterceptorExecutor.java:266)
	at io.servicetalk.http.router.jersey.resources.CancellableResources.lambda$postOioStreams$0(CancellableResources.java:95)
	at org.glassfish.jersey.message.internal.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:55)
	at org.glassfish.jersey.message.internal.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:37)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:242)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:227)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
	at org.glassfish.jersey.server.internal.JsonWithPaddingInterceptor.aroundWriteTo(JsonWithPaddingInterceptor.java:85)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:61)
	... 35 more

Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.internal.inject.Providers checkProviderRuntime
WARNING: A provider io.servicetalk.http.router.jersey.resources.CancellableResources registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.servicetalk.http.router.jersey.resources.CancellableResources will be ignored. 
Mar 08, 2019 10:14:58 PM org.glassfish.jersey.server.ServerRuntime$Responder writeResponse
SEVERE: An I/O error has occurred while writing a response message entity chunk to the container output stream.
java.io.IOException: Already closed TerminalNotification{COMPLETE}
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.processClosed(ConnectablePayloadWriter.java:174)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.verifyOpen(ConnectablePayloadWriter.java:159)
	at io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter.flush(ConnectablePayloadWriter.java:112)
	at io.servicetalk.concurrent.api.internal.ConnectableOutputStream.flush(ConnectableOutputStream.java:64)
	at io.servicetalk.http.router.jersey.DefaultContainerResponseWriter$CopyingOutputStream.flush(DefaultContainerResponseWriter.java:315)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.flush(CommittingOutputStream.java:263)
	at org.glassfish.jersey.server.ChunkedOutput$1.call(ChunkedOutput.java:245)
	at org.glassfish.jersey.server.ChunkedOutput$1.call(ChunkedOutput.java:195)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:219)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:314)
	at org.glassfish.jersey.server.ChunkedOutput.flushQueue(ChunkedOutput.java:195)
	at org.glassfish.jersey.server.ChunkedOutput.setContext(ChunkedOutput.java:374)
	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:687)```

--- a/servicetalk-concurrent-api-internal/src/main/java/io/servicetalk/concurrent/api/internal/ConnectablePayloadWriter.java
+++ b/servicetalk-concurrent-api-internal/src/main/java/io/servicetalk/concurrent/api/internal/ConnectablePayloadWriter.java
@@ -82,16 +82,20 @@ public final class ConnectablePayloadWriter<T> implements PayloadWriter<T> {
 
     @Override
     public void write(final T t) throws IOException {
-        // We don't need to call verifyOpen here because it is checked in both wait* methods.
         for (;;) {
             final long requested = this.requested;
             if (requested > 0) {
                 if (requestedUpdater.compareAndSet(this, requested, requested - 1)) {
                     break;
                 }
-            } else {
+            } else if (requested >= 0) {
                 waitForRequestNDemand();
                 break;
+            } else {
+                // In the event that we have been terminated don't bother trying to wait. More importantly we don't
+                // want to reset the requested state to REQUESTN_ABOUT_TO_PARK because it may have been set to
+                // REQUESTN_TERMINATED.
+                processClosed();
             }
         }
 
@@ -100,6 +104,7 @@ public final class ConnectablePayloadWriter<T> implements PayloadWriter<T> {
             s.onNext(t);
         } catch (final Throwable cause) {
             closed = TerminalNotification.error(cause);
+            requested = REQUESTN_TERMINATED;
             state = State.TERMINATED;
             s.onError(cause);
             throw cause;
@@ -117,16 +122,18 @@ public final class ConnectablePayloadWriter<T> implements PayloadWriter<T> {
         // Set closed before state, because the Subscriber thread depends upon this ordering in the event it needs to
         // terminate the Subscriber.
         if (closedUpdater.compareAndSet(this, null, TerminalNotification.complete())) {
+            // We need to terminate requested or else we may block indefinitely on subsequent calls to write in
+            // waitForRequestNDemand.
+            requested = REQUESTN_TERMINATED;
             for (;;) {
                 final Object currState = state;
-                if (currState == State.TERMINATED || currState == State.CONNECTED) {
-                    break;
-                } else if (currState instanceof Subscriber) {
+                if (currState instanceof Subscriber) {
                     if (stateUpdater.compareAndSet(this, currState, State.TERMINATED)) {
                         ((Subscriber) currState).onComplete();
                         break;
                     }
-                } else if (stateUpdater.compareAndSet(this, currState, State.TERMINATING)) {
+                } else if (currState == State.TERMINATED ||
+                        stateUpdater.compareAndSet(this, currState, State.TERMINATING)) {
                     assert currState != State.WAITING_FOR_CONNECTED;
                     break;
                 }
@@ -273,10 +280,16 @@ public final class ConnectablePayloadWriter<T> implements PayloadWriter<T> {
                                             addWithOverflowProtection(requested, n))) {
                                         break;
                                     }
-                                } else {
-                                    if (requested == REQUESTN_ABOUT_TO_PARK) {
-                                        wakeupWriterThread(n);
+                                } else if (requested == REQUESTN_ABOUT_TO_PARK) {
+                                    // It is possible the writer thread did a GaS to make REQUESTN_ABOUT_TO_PARK visible
+                                    // but if the Get returned some positive demand it will write something and
+                                    // atomically decrement the demand. So we should do a CaS here to be sure we don't
+                                    // overwrite this value and lose demand.
+                                    if (requestedUpdater.compareAndSet(outer, REQUESTN_ABOUT_TO_PARK, n)) {
+                                        tryWakeupWriterThread();
+                                        break;
                                     }
+                                } else {
                                     break;
                                 }
                             }
@@ -324,22 +337,15 @@ public final class ConnectablePayloadWriter<T> implements PayloadWriter<T> {
         }
 
         private void terminateRequestN() {
-            for (;;) {
-                final long requested = outer.requested;
-                if (requested == REQUESTN_ABOUT_TO_PARK) {
-                    wakeupWriterThread(REQUESTN_TERMINATED);
-                    break;
-                } else if (requestedUpdater.compareAndSet(outer, requested, REQUESTN_TERMINATED)) {
-                    break;
-                }
-            }
+            outer.requested = REQUESTN_TERMINATED;
+            tryWakeupWriterThread();
         }
 
-        private void wakeupWriterThread(long requestN) {
+        private void tryWakeupWriterThread() {
             final Thread writerThread = outer.writerThread;
-            assert writerThread != null;
-            outer.requested = requestN; // make the requestN visible to writerThread, before unpark.
-            LockSupport.unpark(writerThread);
+            if (writerThread != null) {
+                LockSupport.unpark(writerThread);
+            }
         }
     }
 

--- a/servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/internal/ConnectableOutputStreamTest.java
+++ b/servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/internal/ConnectableOutputStreamTest.java
@@ -33,6 +33,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.util.Random;
+import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
@@ -44,6 +45,7 @@ import javax.annotation.Nullable;
 
 import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
 import static io.servicetalk.concurrent.api.internal.ConnectablePayloadWriterTest.toRunnable;
+import static io.servicetalk.concurrent.api.internal.ConnectablePayloadWriterTest.verifyCheckedRunnableException;
 import static io.servicetalk.concurrent.internal.DeliberateException.DELIBERATE_EXCEPTION;
 import static io.servicetalk.concurrent.internal.TerminalNotification.complete;
 import static java.lang.Math.max;
@@ -57,6 +59,7 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.fail;
@@ -64,7 +67,6 @@ import static org.junit.rules.ExpectedException.none;
 
 public class ConnectableOutputStreamTest {
     private static final Logger LOGGER = LoggerFactory.getLogger(ConnectableOutputStreamTest.class);
-
     @Rule
     public final Timeout timeout = new ServiceTalkTestTimeout();
     @Rule
@@ -83,6 +85,176 @@ public class ConnectableOutputStreamTest {
     @After
     public void teardown() {
         executorService.shutdown();
+    }
+
+    @Test
+    public void subscribeDeliverDataSynchronously() throws Exception {
+        AtomicReference<Future<?>> futureRef = new AtomicReference<>();
+        toSource(cos.connect().doAfterSubscribe(subscription -> {
+            subscriber.request(1); // request from the TestPublisherSubscriber!
+            // We want to increase the chance that the writer thread has to wait for the Subscriber to become
+            // available, instead of waiting for the requestN demand.
+            CyclicBarrier barrier = new CyclicBarrier(2);
+            futureRef.compareAndSet(null, executorService.submit(toRunnable(() -> {
+                barrier.await();
+                cos.write(1);
+                cos.flush();
+                cos.close();
+            })));
+            try {
+                barrier.await();
+            } catch (InterruptedException | BrokenBarrierException e) {
+                throw new RuntimeException(e);
+            }
+        })).subscribe(subscriber);
+
+        Future<?> f = futureRef.get();
+        assertNotNull(f);
+        f.get();
+        assertThat(subscriber.takeItems(), contains(new byte[]{1}));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+    }
+
+    @Test
+    public void subscribeCloseSynchronously() throws Exception {
+        AtomicReference<Future<?>> futureRef = new AtomicReference<>();
+        toSource(cos.connect().doAfterSubscribe(subscription -> {
+            // We want to increase the chance that the writer thread has to wait for the Subscriber to become
+            // available, instead of waiting for the requestN demand.
+            CyclicBarrier barrier = new CyclicBarrier(2);
+            futureRef.compareAndSet(null, executorService.submit(toRunnable(() -> {
+                barrier.await();
+                cos.close();
+            })));
+            try {
+                barrier.await();
+            } catch (InterruptedException | BrokenBarrierException e) {
+                throw new RuntimeException(e);
+            }
+        })).subscribe(subscriber);
+
+        Future<?> f = futureRef.get();
+        assertNotNull(f);
+        f.get();
+        assertThat(subscriber.takeTerminal(), is(complete()));
+    }
+
+    @Test
+    public void writeAfterCloseShouldThrow() throws IOException {
+        cos.close();
+        expectedException.expect(IOException.class);
+        cos.write(1);
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void multipleWriteAfterCloseShouldThrow() throws Exception {
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cos.write(1);
+            cos.flush();
+            cos.close();
+            cos.write(2);
+            cos.flush();
+        }));
+
+        toSource(cos.connect()).subscribe(subscriber);
+        subscriber.request(2);
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains(new byte[]{1}));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void connectMultipleWriteAfterCloseShouldThrow() throws Exception {
+        toSource(cos.connect()).subscribe(subscriber);
+        subscriber.request(2);
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cos.write(1);
+            cos.flush();
+            cos.close();
+            cos.write(2);
+            cos.flush();
+        }));
+
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains(new byte[]{1}));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void cancelUnblocksWrite() throws Exception {
+        CyclicBarrier afterFlushBarrier = new CyclicBarrier(2);
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cos.write(1);
+            cos.flush();
+            afterFlushBarrier.await();
+            cos.write(2);
+            cos.flush();
+        }));
+
+        toSource(cos.connect()).subscribe(subscriber);
+        subscriber.request(1);
+        afterFlushBarrier.await();
+        subscriber.cancel();
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains(new byte[]{1}));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+        cos.close(); // should be idempotent
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void connectCancelUnblocksWrite() throws Exception {
+        toSource(cos.connect()).subscribe(subscriber);
+        subscriber.cancel();
+        Future<?> f = executorService.submit(toRunnable(() -> cos.write(1)));
+
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), is(empty()));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+        cos.close(); // should be idempotent
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
     }
 
     @Test
@@ -348,7 +520,7 @@ public class ConnectableOutputStreamTest {
     @Test
     public void invalidRequestN() throws IOException {
         AtomicReference<Throwable> failure = new AtomicReference<>();
-        toSource(cos.connect()).subscribe(new PublisherSource.Subscriber<byte[]>() {
+        toSource(cos.connect()).subscribe(new Subscriber<byte[]>() {
             @Override
             public void onSubscribe(final PublisherSource.Subscription s) {
                 s.request(-1);
@@ -377,7 +549,7 @@ public class ConnectableOutputStreamTest {
     @Test
     public void onNextThrows() throws IOException {
         AtomicReference<Throwable> failure = new AtomicReference<>();
-        toSource(cos.connect()).subscribe(new PublisherSource.Subscriber<byte[]>() {
+        toSource(cos.connect()).subscribe(new Subscriber<byte[]>() {
             @Override
             public void onSubscribe(final PublisherSource.Subscription s) {
                 s.request(1);
@@ -449,8 +621,7 @@ public class ConnectableOutputStreamTest {
             f.get();
             fail();
         } catch (ExecutionException e) {
-            assertThat(e.getCause(), is(instanceOf(RuntimeException.class)));
-            assertThat(e.getCause().getCause(), is(instanceOf(IOException.class)));
+            verifyCheckedRunnableException(e, IOException.class);
         }
         assertThat(subscriber.takeError(), is(instanceOf(IllegalArgumentException.class)));
     }
@@ -471,8 +642,7 @@ public class ConnectableOutputStreamTest {
             f.get();
             fail();
         } catch (ExecutionException e) {
-            assertThat(e.getCause(), is(instanceOf(RuntimeException.class)));
-            assertThat(e.getCause().getCause(), is(instanceOf(IOException.class)));
+            verifyCheckedRunnableException(e, IOException.class);
         }
         assertThat(subscriber.takeError(), is(instanceOf(IllegalArgumentException.class)));
     }

--- a/servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/internal/ConnectablePayloadWriterTest.java
+++ b/servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/internal/ConnectablePayloadWriterTest.java
@@ -15,7 +15,8 @@
  */
 package io.servicetalk.concurrent.api.internal;
 
-import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+import io.servicetalk.concurrent.PublisherSource.Subscription;
 import io.servicetalk.concurrent.api.Publisher;
 import io.servicetalk.concurrent.api.TestPublisherSubscriber;
 import io.servicetalk.concurrent.internal.ServiceTalkTestTimeout;
@@ -31,6 +32,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.util.Random;
+import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
@@ -55,6 +57,7 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.fail;
@@ -65,7 +68,8 @@ public class ConnectablePayloadWriterTest {
     public final Timeout timeout = new ServiceTalkTestTimeout();
     @Rule
     public final ExpectedException expectedException = ExpectedException.none();
-    public final TestPublisherSubscriber<String> subscriber = new TestPublisherSubscriber<>();
+
+    private final TestPublisherSubscriber<String> subscriber = new TestPublisherSubscriber<>();
     private ConnectablePayloadWriter<String> cpw;
     private ExecutorService executorService;
 
@@ -78,6 +82,176 @@ public class ConnectablePayloadWriterTest {
     @After
     public void teardown() {
         executorService.shutdown();
+    }
+
+    @Test
+    public void subscribeDeliverDataSynchronously() throws Exception {
+        AtomicReference<Future<?>> futureRef = new AtomicReference<>();
+        toSource(cpw.connect().doAfterSubscribe(subscription -> {
+            subscriber.request(1); // request from the TestPublisherSubscriber!
+            // We want to increase the chance that the writer thread has to wait for the Subscriber to become
+            // available, instead of waiting for the requestN demand.
+            CyclicBarrier barrier = new CyclicBarrier(2);
+            futureRef.compareAndSet(null, executorService.submit(toRunnable(() -> {
+                barrier.await();
+                cpw.write("foo");
+                cpw.flush();
+                cpw.close();
+            })));
+            try {
+                barrier.await();
+            } catch (InterruptedException | BrokenBarrierException e) {
+                throw new RuntimeException(e);
+            }
+        })).subscribe(subscriber);
+
+        Future<?> f = futureRef.get();
+        assertNotNull(f);
+        f.get();
+        assertThat(subscriber.takeItems(), contains("foo"));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+    }
+
+    @Test
+    public void subscribeCloseSynchronously() throws Exception {
+        AtomicReference<Future<?>> futureRef = new AtomicReference<>();
+        toSource(cpw.connect().doAfterSubscribe(subscription -> {
+            // We want to increase the chance that the writer thread has to wait for the Subscriber to become
+            // available, instead of waiting for the requestN demand.
+            CyclicBarrier barrier = new CyclicBarrier(2);
+            futureRef.compareAndSet(null, executorService.submit(toRunnable(() -> {
+                barrier.await();
+                cpw.close();
+            })));
+            try {
+                barrier.await();
+            } catch (InterruptedException | BrokenBarrierException e) {
+                throw new RuntimeException(e);
+            }
+        })).subscribe(subscriber);
+
+        Future<?> f = futureRef.get();
+        assertNotNull(f);
+        f.get();
+        assertThat(subscriber.takeTerminal(), is(complete()));
+    }
+
+    @Test
+    public void writeAfterCloseShouldThrow() throws IOException {
+        cpw.close();
+        expectedException.expect(IOException.class);
+        cpw.write("foo");
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void multipleWriteAfterCloseShouldThrow() throws Exception {
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cpw.write("foo");
+            cpw.flush();
+            cpw.close();
+            cpw.write("bar");
+            cpw.flush();
+        }));
+
+        toSource(cpw.connect()).subscribe(subscriber);
+        subscriber.request(2);
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains("foo"));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void connectMultipleWriteAfterCloseShouldThrow() throws Exception {
+        toSource(cpw.connect()).subscribe(subscriber);
+        subscriber.request(2);
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cpw.write("foo");
+            cpw.flush();
+            cpw.close();
+            cpw.write("bar");
+            cpw.flush();
+        }));
+
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains("foo"));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void cancelUnblocksWrite() throws Exception {
+        CyclicBarrier afterFlushBarrier = new CyclicBarrier(2);
+        Future<?> f = executorService.submit(toRunnable(() -> {
+            cpw.write("foo");
+            cpw.flush();
+            afterFlushBarrier.await();
+            cpw.write("bar");
+            cpw.flush();
+        }));
+
+        toSource(cpw.connect()).subscribe(subscriber);
+        subscriber.request(1);
+        afterFlushBarrier.await();
+        subscriber.cancel();
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), contains("foo"));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+        cpw.close(); // should be idempotent
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
+    }
+
+    @Test
+    public void connectCancelUnblocksWrite() throws Exception {
+        toSource(cpw.connect()).subscribe(subscriber);
+        subscriber.cancel();
+        Future<?> f = executorService.submit(toRunnable(() -> cpw.write("foo")));
+
+        try {
+            f.get();
+            fail();
+        } catch (ExecutionException e) {
+            verifyCheckedRunnableException(e, IOException.class);
+        }
+
+        assertThat(subscriber.takeItems(), is(empty()));
+        assertThat(subscriber.takeTerminal(), is(complete()));
+        cpw.close(); // should be idempotent
+
+        // Make sure the Subscription thread isn't blocked.
+        subscriber.request(1);
+        subscriber.cancel();
     }
 
     @Test
@@ -108,9 +282,9 @@ public class ConnectablePayloadWriterTest {
         CountDownLatch onSubscribe = new CountDownLatch(1);
         CountDownLatch onComplete = new CountDownLatch(1);
         AtomicReference<Throwable> errorRef = new AtomicReference<>();
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 s.request(-1);
                 onSubscribe.countDown();
             }
@@ -141,9 +315,9 @@ public class ConnectablePayloadWriterTest {
     public void multipleConnectWhileEmittingShouldFailConnect() throws Exception {
         CountDownLatch onNext = new CountDownLatch(1);
         CountDownLatch onComplete = new CountDownLatch(1);
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 s.request(1);
             }
 
@@ -174,9 +348,9 @@ public class ConnectablePayloadWriterTest {
     public void multipleConnectWhileSubscribedShouldFailConnect() throws Exception {
         CountDownLatch onSubscribe = new CountDownLatch(1);
         CountDownLatch onComplete = new CountDownLatch(1);
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 onSubscribe.countDown();
             }
 
@@ -204,9 +378,9 @@ public class ConnectablePayloadWriterTest {
     public void multipleConnectWhileSubscriberFailedShouldFailConnect() throws Exception {
         CountDownLatch onError = new CountDownLatch(1);
         CountDownLatch onComplete = new CountDownLatch(1);
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 s.request(1);
             }
 
@@ -343,9 +517,9 @@ public class ConnectablePayloadWriterTest {
     @Test
     public void invalidRequestN() throws IOException {
         AtomicReference<Throwable> failure = new AtomicReference<>();
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 s.request(-1);
             }
 
@@ -372,9 +546,9 @@ public class ConnectablePayloadWriterTest {
     @Test
     public void onNextThrows() throws IOException {
         AtomicReference<Throwable> failure = new AtomicReference<>();
-        toSource(cpw.connect()).subscribe(new PublisherSource.Subscriber<String>() {
+        toSource(cpw.connect()).subscribe(new Subscriber<String>() {
             @Override
-            public void onSubscribe(final PublisherSource.Subscription s) {
+            public void onSubscribe(final Subscription s) {
                 s.request(1);
             }
 
@@ -444,8 +618,7 @@ public class ConnectablePayloadWriterTest {
             f.get();
             fail();
         } catch (ExecutionException e) {
-            assertThat(e.getCause(), is(instanceOf(RuntimeException.class)));
-            assertThat(e.getCause().getCause(), is(instanceOf(IOException.class)));
+            verifyCheckedRunnableException(e, IOException.class);
         }
         assertThat(subscriber.takeError(), is(instanceOf(IllegalArgumentException.class)));
     }
@@ -466,8 +639,7 @@ public class ConnectablePayloadWriterTest {
             f.get();
             fail();
         } catch (ExecutionException e) {
-            assertThat(e.getCause(), is(instanceOf(RuntimeException.class)));
-            assertThat(e.getCause().getCause(), is(instanceOf(IOException.class)));
+            verifyCheckedRunnableException(e, IOException.class);
         }
         assertThat(subscriber.takeError(), is(instanceOf(IllegalArgumentException.class)));
     }
@@ -532,13 +704,13 @@ public class ConnectablePayloadWriterTest {
         final Thread consumerThread = new Thread(() -> {
             try {
                 final CountDownLatch consumerDone = new CountDownLatch(1);
-                toSource(pub).subscribe(new PublisherSource.Subscriber<String>() {
+                toSource(pub).subscribe(new Subscriber<String>() {
                     @Nullable
-                    private PublisherSource.Subscription sub;
+                    private Subscription sub;
                     private int writeIndex;
 
                     @Override
-                    public void onSubscribe(final PublisherSource.Subscription s) {
+                    public void onSubscribe(final Subscription s) {
                         sub = s;
                         sub.request(1);
                     }
@@ -593,5 +765,10 @@ public class ConnectablePayloadWriterTest {
     @FunctionalInterface
     interface CheckedRunnable {
         void doWork() throws Exception;
+    }
+
+    static void verifyCheckedRunnableException(ExecutionException e, Class<? extends Throwable> clazz) {
+        assertThat(e.getCause(), is(instanceOf(RuntimeException.class))); // this is from toRunnable
+        assertThat(e.getCause().getCause(), is(instanceOf(clazz)));
     }
 }

--- a/servicetalk-http-router-jersey/src/testFixtures/java/io/servicetalk/http/router/jersey/CancellationTest.java
+++ b/servicetalk-http-router-jersey/src/testFixtures/java/io/servicetalk/http/router/jersey/CancellationTest.java
@@ -33,7 +33,6 @@ import io.servicetalk.transport.api.ExecutionContext;
 import io.servicetalk.transport.api.IoExecutor;
 
 import org.junit.Before;
-import org.junit.ClassRule;
 import org.junit.Rule;
 import org.junit.Test;
 import org.mockito.Mock;
@@ -51,7 +50,6 @@ import javax.ws.rs.core.Application;
 import static io.servicetalk.buffer.netty.BufferAllocators.DEFAULT_ALLOCATOR;
 import static io.servicetalk.concurrent.api.Publisher.just;
 import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
-import static io.servicetalk.concurrent.internal.ServiceTalkTestTimeout.DEFAULT_TIMEOUT_SECONDS;
 import static io.servicetalk.http.api.HttpExecutionStrategies.defaultStrategy;
 import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_TYPE;
 import static io.servicetalk.http.api.HttpHeaderValues.TEXT_PLAIN;
@@ -63,7 +61,6 @@ import static java.net.InetSocketAddress.createUnresolved;
 import static java.util.Collections.singleton;
 import static java.util.Collections.singletonMap;
 import static java.util.concurrent.TimeUnit.DAYS;
-import static java.util.concurrent.TimeUnit.SECONDS;
 import static java.util.function.Function.identity;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.nullValue;
@@ -92,8 +89,8 @@ public class CancellationTest {
 
     private static final CharSequence TEST_DATA = newLargePayload();
 
-    @ClassRule
-    public static final ExecutorRule EXEC = ExecutorRule.newRule();
+    @Rule
+    public final ExecutorRule execRule = ExecutorRule.newRule();
 
     @Rule
     public final MockitoRule rule = MockitoJUnit.rule();
@@ -105,7 +102,7 @@ public class CancellationTest {
     private HttpServiceContext ctx;
 
     @Mock
-    private Executor exec;
+    private Executor execMock;
 
     private CancellableResources cancellableResources;
     private StreamingHttpService jerseyRouter;
@@ -116,14 +113,14 @@ public class CancellationTest {
         when(ctx.executionContext()).thenReturn(execCtx);
         when(ctx.localAddress()).thenReturn(createUnresolved("localhost", 8080));
         when(execCtx.bufferAllocator()).thenReturn(DEFAULT_ALLOCATOR);
-        when(execCtx.executor()).thenReturn(exec);
+        when(execCtx.executor()).thenReturn(execMock);
         when(execCtx.ioExecutor()).thenReturn(mock(IoExecutor.class));
 
         cancellableResources = new CancellableResources();
 
         jerseyRouter = new HttpJerseyRouterBuilder()
                 .routeExecutionStrategyFactory(asFactory(
-                        singletonMap("test", defaultStrategy(EXEC.executor()))))
+                        singletonMap("test", defaultStrategy(execRule.executor()))))
                 .build(new Application() {
                     @Override
                     public Set<Object> getSingletons() {
@@ -137,7 +134,7 @@ public class CancellationTest {
     @Test
     public void cancelSuspended() throws Exception {
         final TestCancellable cancellable = new TestCancellable();
-        when(exec.schedule(any(Runnable.class), eq(7L), eq(DAYS))).thenReturn(cancellable);
+        when(execMock.schedule(any(Runnable.class), eq(7L), eq(DAYS))).thenReturn(cancellable);
 
         testCancelResponseSingle(get("/suspended"), false);
         assertThat(cancellable.cancelled, is(true));
@@ -174,20 +171,20 @@ public class CancellationTest {
     public void cancelSse() throws Exception {
         doAnswer(invocation -> {
             Object[] args = invocation.getArguments();
-            return EXEC.executor().schedule((Runnable) args[0], (long) args[1], (TimeUnit) args[2]);
-        }).when(exec).schedule(any(Runnable.class), anyLong(), any(TimeUnit.class));
+            return execRule.executor().schedule((Runnable) args[0], (long) args[1], (TimeUnit) args[2]);
+        }).when(execMock).schedule(any(Runnable.class), anyLong(), any(TimeUnit.class));
 
         // Initial SSE request succeeds
         testCancelResponsePayload(get("/sse"));
 
         // Payload cancellation closes the SSE sink
-        assertThat(cancellableResources.sseSinkClosedLatch.await(DEFAULT_TIMEOUT_SECONDS, SECONDS), is(true));
+        cancellableResources.sseSinkClosedLatch.await();
     }
 
     private void testCancelResponsePayload(final StreamingHttpRequest req) throws Exception {
         // The handler method uses OutputStream APIs which are blocking. So we need to call handle and subscribe on
         // different threads because the write operation will block on the Subscriber creating requestN demand.
-        Single<StreamingHttpResponse> respSingle = EXEC.executor().submit(() ->
+        Single<StreamingHttpResponse> respSingle = execRule.executor().submit(() ->
                     jerseyRouter.handle(ctx, req, HTTP_REQ_RES_FACTORY))
                 .flatMap(identity());
         Future<StreamingHttpResponse> respFuture = respSingle.toFuture();
@@ -221,7 +218,7 @@ public class CancellationTest {
         // The handler method uses OutputStream APIs which are blocking. So we need to call handle and subscribe on
         // different threads because the write operation will block on the Subscriber creating requestN demand.
         if (enableOffload) {
-            Single<StreamingHttpResponse> respSingle = EXEC.executor().submit(() ->
+            Single<StreamingHttpResponse> respSingle = execRule.executor().submit(() ->
                     jerseyRouter.handle(ctx, req, HTTP_REQ_RES_FACTORY)
             ).flatMap(identity())
              .doBeforeError(errorRef::set)

